onnx/shape_inference/implementation.cc

```diff
@@ -688,8 +688,28 @@ std::vector<const TypeProto*> GraphInferencerImpl::doInferencing(
   SymbolTable* symbolTable = getSymbolTable();
   int numInputs = int(inputTypes.size());
 
-  if (g_->input_size() != numInputs) {
-    fail_shape_inference("Graph has ", g_->input_size(), " inputs but ", numInputs, " were provided");
+  if (getIRVersion() >= 4) {
+    if (g_->input_size() != numInputs) {
+      fail_shape_inference("Graph has ", g_->input_size(), " inputs but ", numInputs, " were provided");
+    }
+  } else {
+    std::unordered_set<std::string> initializerNameSet;
+    for (const auto& tp : g_->initializer()) {
+      initializerNameSet.insert(tp.name());
+    }
+    // IR < 4 requires all initializers to be optional inputs
+    // So the number of graph input can be larger than the number of node input 
+    if (numInputs > g_->input_size()) {
+      fail_shape_inference("Graph has ", g_->input_size(), " inputs but ", numInputs, " were provided.",
+        "The number of graph input cannot be smaller than the number of node input" );
+    } else if (numInputs < g_->input_size()) {
+      // Since the number is different, further check whether the additional input is in initializers
+      for (int i = numInputs; i < g_->input_size(); ++i) {
```

Comment:
You also need to check that the initial inputs do not have initializers, if you are disallowing them. (see first comment)

Reply:
Good point. I added the check (initial inputs do not have initializers) here and there (your first comment). Thanks!